FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,124 @@
+
+Every `XarrayIndex` subclass must at least implement a method that takes label-based indexers as argument and that returns the corresponding position-based indexers. Whether or not a `XarrayIndex` subclass supports each of the features listed above should be either declared explicitly via a common API or left to the implementation.
+
+There are potentially other methods that an `XarrayIndex` subclass must/should/may implement, like Xarray coordinate getters (see [Section 2.2.4](#224_Implicit_coodinates)) or `serialize()`/`deserialize()` methods. `XarrayIndex` API may be expanded in the future.
+
+Xarray should provide a minimal set of built-in index wrappers (this could be reduced to the indexes currently supported in Xarray, i.e., `pandas.Index` and `pandas.MultiIndex`). Other index wrappers may be implemented in 3rd-party libraries (recommended). The `XarrayIndex` base class should be part of Xarray's public API.
+
+#### 2.1.1 Index discoverability
+
+For better discoverability of Xarray-compatible indexes, Xarray could provide some mechanism to register new index wrappers, e.g., something like [xoak's `IndexRegistry`](https://xoak.readthedocs.io/en/latest/_api_generated/xoak.IndexRegistry.html#xoak.IndexRegistry).
+
+### 2.2 Explicit vs. implicit index creation
+
+#### 2.2.1 Dataset/DataArray's `set_index` method
+
+New indexes can be built from an existing set of coordinates or variables in a Dataset/DataArray using the `.set_index()` method.
+
+TODO: describe API updates needed to provide the kind of index that we want to build from a set of coordinates.
+
+#### 2.2.2 Dataset/DataArray's `indexes` constructor argument
+
+TODO
+
+#### 2.2.3 Implicit default indexes
+
+In general explicit index creation should be preferred over implicit index creation. However, there is a majority of cases where basic `pandas.Index` objects could be built and used as indexes for 1-dimensional coordinates. For convenience, Xarray should automatically build such indexes for the coordinates where no index has been explicitly assigned in the Dataset/DataArray constructor or when indexes have been reset / dropped.
+
+For which coordinates?
+
+- A. only 1D coordinates with a name matching their dimension name
+- B. all 1D coordinates
+
+When to create it?
+
+- A. each time when a new Dataset/DataArray is created
+- B. only when we need it (i.e., when calling `.sel()` or `indexes`)
+
+#### 2.2.4 Implicit coordinates
+
+If we want to strictly follow the data model defined in [Section 1](#1_Data_Model), we need to ensure that at least one coordinate exists for each index (generally we need to ensure that the data indexed is properly exposed as one or more coordinates), especially when index objects are given directly to the Dataset/DataArray constructor ([Section 2.2.2](#222_DatasetDataArrays_indexes_constructor_argument)). This means that every `XarrayIndex` subclass must implement a method to get or generate one or more Xarray coordinates.
```

Comment:
Another option would be to only validate that the corresponding coordinate exist when `indexes` are passed, and not to do implicit creation of coordinates from an index. I think this is viable given that we currently don't support explicitly passing an index.

Reply:
Hmm currently we can explicitly pass a multi-index as a coordinate:

```python
>>> midx = pd.MultiIndex.from_arrays([['a', 'b'], [0, 1]], names=['lvl1', 'lvl2'])
>>> xr.DataArray([1.0, 2.0], dims='x', coords={'x': midx})
```

Virtual coordinates are created for each level of the multi-index. How should we handle this case after the refactoring? We could silently ignore the levels and simply consider `midx` as an array (i.e., `midx.to_numpy()`). Or we could raise an error...

More generally, I thought that explicitly passing index objects could still be useful since those objects may be costly to build, but I need to think more about whether it's worth the trouble.

Reply:
> Yes, I would still like to support this syntax when a MultiIndex is passed in coords, but maybe by unpacking as real rather than "virtual" coordinates. This could directly make coordinates with the names lvl1 and lvl2.

Hmm shouldn't we take the opportunity of this refactoring to change this behavior? This looks already pretty confusing to me, and it will be even more if it creates new variable objects that do not have the name passed as key to `coords`. Even though it would be a breaking change, a more predictable behavior would be that any value in the sequence or mapping passed to `coords` would be treated as an array that can be wrapped into an xarray variable, i.e.,

```python
>>> xr.DataArray([1.0, 2.0], dims='x', coords={'x': midx})
<xarray.DataArray (x: 2)>
array([1., 2.])
Coordinates:
    x        (x) object ('a', 0) ('b', 1)
```

If we really want to reuse an existing index object in a Dataset/DataArray (which I assume is not a *very* common case), we could maybe do so via some helper function, e.g.,

```python
>>> coords_dict, index_dict = create_coords_from_index(midx, dims='x')

>>> xr.DataArray([1.0, 2.0], dims='x', coords=coords_dict, indexes=index_dict)
<xarray.DataArray (x: 2)>
array([1., 2.])
Coordinates:
  * lvl1     (x) object 'a' 'b'
  * lvl2     (x) int64 0 1
Indexes:
    lvl1, lvl2   <MultiIndex coords=(lvl1, lvl2) dims=x>
```